### PR TITLE
Main window toggle button

### PR DESCRIPTION
**Pull request by meadowsjared**
Adds a toggle to the main icon, so if you left-click on it, it toggles the `mainWindow` status of that window.
Additionaly, it should alter the icon of that `mainWindow` to make it clear to the user which window is the `mainWindow`.

I'm open to suggestion on what the highlighted icon should look like.  Feel free to suggest/and/or provide one, as this is your app, and I don't want to overstep. Cheers!🥂

### CODE REVIEW

**Comment by meadowsjared** [comment]
I've got it working, but it's looking like Chrome doesn't allow me to set a different Icon per window, so the way it's working right now is, when you click it, it updates the icon for ALL windows.... 😞 I'll look into it more later, hopefully there's a workaround.

**Comment by meadowsjared** [comment]
well it IS technically possible to set the icon for one window, but it's a pain in the neck, you have to set it per-tab.  Which means, I have to update the tab when it gets moved, created, etc.... very dumb.  They should allow you to set it by windowId like firefox allows... smh...

**Comment by meadowsjared** [comment]
I'm still doing more testing, I got it working pretty smooth but I have one last error that I came across saying "no tab with id 3425" but I'm not sure what's causing it yet. 🤷‍♂️  I'll check it out later.

**Comment by meadowsjared** [comment]
Also, you can delete the `icon.pdn` file, I only included it in case you wanted to tweak it.  But feel free to delete it if you have no use for it.

**Comment by meadowsjared** [comment]
ok it should be all good now, feel free to criticize if you don't like how I had to get it working, it wasn't fun. 👎

File: manifest.json [hunk 1]
```diff
@@ -1,9 +1,13 @@
 {
   "name": "Force New Tabs to Main Window",
   "description": "Designate a single main window where new tabs should open",
-  "version": "1.0.4",
+  "version": "1.1.0",
   "manifest_version": 3,
   "permissions": ["tabs", "storage"],
+  "homepage_url": "https://chrome.google.com/webstore/detail/force-new-tabs-to-main-wi/kgojjafnbajjomijiceiefakpphcicen",
```
Comment (timtrinidad) [inline manifest.json:7]: Thanks for setting this!

File: src/background.js [hunk 2]
```diff
@@ -1,3 +1,111 @@
+// only update only the mainWindowIdOld and the mainWindowId
```
Comment (timtrinidad) [inline src/background.js]: Nit - use docbloc comments format and more descriptive comment
```
/**
 * Update the icons of the specified window ID to the "main window" icon" (and remove the existing icon)
 */
```

Comment (timtrinidad) [reply src/background.js]: Also - can you add a comment specifying why for something that looks like it updates the window, needs to set the icon for each individual tab? 

What happens if a new tab is added to one of these windows?

**Review by timtrinidad — CHANGES_REQUESTED** [review]

**Review by timtrinidad — CHANGES_REQUESTED** [review]
I like where you're going with this, however, like you said it's a lot of delicate workarounds to get it working. I also somehow ended up in a state where the icons for the same window were switching between tabs.


Some thoughts:
* Are people switching the "main window" so often that it warrants complicating the extension code this much?
* Almost all of my extensions show a menu rather than executing an action when you click on the button. Would this be cleaner if we simply showed a dropdown with a check-able menu item like ":check: Main Window" or something like that?

File: manifest.json (hunk 1 above)
Comment (meadowsjared) [reply manifest.json:7]: no prob! after hijacking the left click action, it was necessary to add the `homepage_url` so clicking the context menu would still take you to the same place.

File: src/background.js (hunk 2 above)
Comment (meadowsjared) [reply src/background.js]: - ok, I will do that next time I can work on it
- my intention (assuming all is working correctly) is that if you click it, it simply sets your "MainWindow" (the same as if you right-click, hit options, wait for it to load, then select the window from the drop down and hit save).  The difference is, this acts as a quick-toggle (as well as an easy-to-identify indicator).  So clicking it once sets the "MainWindow" to the window you just clicked on.  And clicking it a second time will un-set it, so now MainWindow is set to 0 (aka turned off).
- If a new tab is added to one of these windows it should behave the same as before, (get moved to the main window if applicable), however if the window the tab is added to is the main window, then the icon should be set to the X icon.
- Basically my intention is that, to the user, when the user clicks it on a certain window, no matter what tab you're on, the icon for that window will have the special X icon.  So it shouldn't really follow any tabs, the icon should be consistent per-window if that makes sense.
- if a tab gets drug away (detached) from that window to another window that's not main, it will technically lose the special icon, because it's no longer on the mainWindow.  But to the user, it will just appear to adopt the icon of the destination window.

**Comment by meadowsjared** [comment]
I think it's less about them switching it so often, and more about making it clear, which window is selected.  And also, in using it, I do really like being able to just click on it, rather than going though the options window, it takes many more clicks that way.
I see a fair number of extensions that use the main button to "activate" the extension.  Take for example this extension:
[Click to Remove Element](https://chrome.google.com/webstore/detail/click-to-remove-element/jcgpghgjhhahcefnfpbncdmhhddedhnk) 
Left-clicking it's main icon "activates it" just like I was proposing for this one, assuming you're ok with that.

- regarding the weird state you found where the icon was changing per-tab (not how I designed it to work), if you can describe to me how I can reproduce that behavior, it'd be very helpful in me troubleshooting how to fix it.  Then again, it's possible I just need to load up this branch and test it some more.

Also, I'm running windows, are you running windows as well? Maybe there's a different behavior due to your operating system?
That could be the source of the weird behavior you discovered 🤔

Of course, this is all assuming I can make it work smoothly.... I'll see what I can do. 🤞

**Comment by jaredmeadowsLB** [comment]
my initial testing looks like it's working as intended (the icon stays constant per-window), if you can tell me what OS you're running on, that would help a TON.  So I can replicate your situation.

**Comment by meadowsjared** [comment]
my initial testing looks like it's working as intended (the icon stays constant per-window), if you can tell me what OS you're running on, that would help a TON. So I can replicate and resolve what's happening with your testing.

**Comment by timtrinidad** [comment]
I am on MacOS Big Sur. IIRC I "cmd-click"ed on a link on one tab to open
that link on a new tab, and that new tab had the incorrect icon.

File: src/background.js [hunk 3]
```diff
@@ -1,3 +1,111 @@
...
+chrome.tabs.onUpdated.addListener(async (tabId, changeInfo, tab) => {
+  if (tab.active && changeInfo.status === 'complete') {
+    chrome.storage.sync.get(
+      {
+        mainWindowId: 0,
+      },
+      ({ mainWindowId }) => {
+        chrome.action.setIcon({
+          path:
+            tab.windowId === mainWindowId ? 'src/icon_x.png' : 'src/icon.png', // set the icon for only for MainWindow
```
Comment (timtrinidad) [inline src/background.js]: I will also think on alternatives for the icon. In my mind, "x" would mean "not", so it's counterintuitive to me that the window marked as "x" is the main window. Perhaps a green or grey dot?

Comment (meadowsjared) [reply src/background.js]: oh I agree, I was attempting to just put something up, but am fully in support of replacing it with whatever you desire.
The reason I initially put an X was, based on the original arrow, I took that to mean, "clicking the arrow makes this the mainWindow".  Therefore the X meant "clicking the X makes this NOT the main window".  But you're right, I don't think an X is the answer.  My initial take on it was I made an icon that was the window border, but filled in blue.  But looking at it, it made it hard to understand what was being shown, it just looked like a box.  I'll give the green or grey dot a shot that seems like it may be better.  The thought process is, it should convey that "this window is Active".  Thank you for the feedback!

Also thanks for explaining the "cmd-click" thing! 🎉  I WAS able to replicate that, and will patch that issue (when I have time ofc).

Also for clarity, I did not technically WANT to code things the way I did.  I could have just updated all tabs, every time a tab is focused.  However in doing so, if I have say.... 100 or more tabs open.  It can start acting laggy if it tries to update all the tabs every time, so the way it's coded right now is, it only updates the icons when it's needed.  Once it's done, I'll see if I can clean up the code at all, or abstract things a little to make it more legible.

Comment (timtrinidad) [reply src/background.js]: This is where a dropdown may help. It still is more clicks than just seeing the icon, but displaying an indicator only when you click (e.g. a checkbox in a dropdown) could simplify the logic a lot (and still provide a convenience over having to open the extension options to see which window is the main window).

File: src/background.js (hunk 2 above)
Comment (meadowsjared) [reply src/background.js]: I believe I've added the comments you requested

File: src/background.js (hunk 3 above)
Comment (meadowsjared) [reply src/background.js]: I believe I've fixed the issue with the icon not working properly if a link it middle-clicked (aka command-clicked) (previously it would cause the tab to get created, yet not be focused, which caused it to not get assigned, however now it is properly assigned).

I don't believe the menu would help as I believe it would still be global, however either way, it should be working now, but please let me know if you can find any issues with it.

**Comment by meadowsjared** [comment]
the activated version of the icon is now highlighted around the edges in green.

If you'd like to see it, you can change `icon_highlighted.png` to `icon_dot.png` on line 20 to see how that turned out.  I tried it, but it looked a little bad imho, but you're welcome to use it if you prefer it.  It's possible it looks better on yours as I think part of why it looks so bad on mine is I'm only using a 1080p monitor, so my icon looks like this at my resolution:
blurry box
you can see that in the above image, you can't even see the 3 dots, which made this much harder to not just make it look like a mess.  In the highlighted version I even made the lines much bolder, but you don't notice it very much because it's so small.

Lastly, don't let me forget we should delete whichever icon you don't like as much, as well as the `image.pdn` file (I only included that in case you wanted the source files I used to create the images.)

**Comment by jaredmeadowsLB** [comment]
@timtrinidad please by all means feel free to replace the icon I whipped up with something you like more.  That's why I included multiple options, so you can test them out, plus I refactored the code so the filename is only written once for simplicity. ☺

**Comment by meadowsjared** [comment]
@timtrinidad please by all means feel free to replace the icon I whipped up with something you like more. That's why I included multiple options, so you can test them out, plus I refactored the code so the filename is only written once for simplicity. ☺

**Review by timtrinidad — APPROVED** [review]
Looks much better, and is working great on my end. Approved!

**Comment by meadowsjared** [comment]
Thank you @timtrinidad glad to hear it, I've been testing it this last two weeks and haven't found any issues with it.  Now that this is working as intended, I'll get started on #6! (when I have time ofc) 🎉
